### PR TITLE
[Test] Smarter enumerating features and raise exception when not supported

### PR DESCRIPTION
**Pull request by squarefk**
Related issue = #

### CODE REVIEW

File: python/taichi/testing.py [hunk 1]
```diff
@@ -109,13 +127,18 @@ def wrapped(*args, **kwargs):
                 skip = False
                 current_options = copy.deepcopy(options)
                 for i, feature in enumerate(_test_features):
```
Comment (k-ye) [inline python/taichi/testing.py:129]: nit: `_test_features[1:]`

File: python/taichi/testing.py [hunk 2]
```diff
@@ -55,9 +55,27 @@ def make_temp_file(*args, **kwargs):
...
 
-# To test full features set:
-# _test_features = {"packed": [True, False], "dynamic_index": [True, False]}
-_test_features = {"dynamic_index": [True, False]}
+class TestValue:
+    def __init__(self, value, required_extensions):
+        self.value = value
+        self.required_extensions = required_extensions
+
+    def get_value(self):
```
Comment (k-ye) [inline python/taichi/testing.py:63]: @property
def value(self):
  return self._value

File: python/taichi/testing.py [hunk 3]
```diff
@@ -55,9 +55,27 @@ def make_temp_file(*args, **kwargs):
...
+        return self.value
+
+    def get_required_extensions(self):
+        return self.required_extensions
+
```
Comment (k-ye) [inline python/taichi/testing.py:68]: ditto

**Review by k-ye — APPROVED** [review]
LGTM!

File: python/taichi/testing.py (hunk 3 above)
Comment (squarefk) [reply python/taichi/testing.py]: https://github.com/taichi-dev/taichi/pull/2679/commits/3eb3782c0d56507f812ced7afe89eeea67a3c22b

File: python/taichi/testing.py (hunk 2 above)
Comment (squarefk) [reply python/taichi/testing.py]: https://github.com/taichi-dev/taichi/pull/2679/commits/3eb3782c0d56507f812ced7afe89eeea67a3c22b

File: python/taichi/testing.py (hunk 1 above)
Comment (squarefk) [reply python/taichi/testing.py]: https://github.com/taichi-dev/taichi/pull/2679/commits/3eb3782c0d56507f812ced7afe89eeea67a3c22b

**Pull request merged**